packages/authentication/test/acceptance/basic-auth.ts

```diff
@@ -0,0 +1,206 @@
+    app = new Application();
+    app.bind('application.name').to('SequenceApp');
+  }
+
+  function givenControllerInApp() {
+    const apispec = givenOpenApiSpec()
+      .withOperation('get', '/whoAmI', {
+        'x-operation-name': 'whoAmI',
+        responses: {
+          '200': {
+            type: 'string',
+          },
+        },
+      })
+      .build();
+
+    @api(apispec)
+    class MyController {
+      constructor(@inject('authentication.user') private user: UserProfile) {}
+
+      @authenticate('BasicStrategy')
+      async whoAmI() : Promise<string> {
+        return this.user.id;
+      }
+    }
+    app.controller(MyController);
+  }
+
+  function givenAuthenticatedSequence() {
+    class MySequence {
+      constructor(
+        @inject('findRoute') protected findRoute: FindRoute,
+        @inject('getFromContext') protected getFromContext: getFromContext,
+        @inject('invokeMethod') protected invoke: InvokeMethod,
+        @inject('bindElement') protected bindElement: bindElement,
+      ) {}
+
+      async run(req: ParsedRequest, res: ServerResponse) {
+        try {
+          const {
+            controller,
+            methodName,
+            spec: routeSpec,
+            pathParams,
+          } = this.findRoute(req);
+
+          // Resolve authenticate() from AuthenticationProvider
+          const authenticate: AuthenticateFn =
+            await this.getFromContext(BindingKeys.Authentication.PROVIDER);
+
+          // Authenticate
+          const user: UserProfile = await authenticate(req);
+
+          // User is expected to be returned or an exception should be thrown
+          if (user) this.bindElement('authentication.user', user);
+          else throw new HttpErrors.InternalServerError('auth error');
+
+          // Authentication successful, proceed to invoke controller
+          const args = await parseOperationArgs(req, routeSpec, pathParams);
+          const result = await this.invoke(controller, methodName, args);
+          writeResultToResponse(res, result);
+        } catch (err) {
+          this.sendError(res, req, err);
+          return;
+        }
+      }
+      sendError(res: ServerResponse, req: ServerRequest, err: HttpError) {
+        const statusCode = err.statusCode || err.status || 500;
+        res.statusCode = statusCode;
+        res.end(err.message);
+      }
+    }
+    // bind user defined sequence
+    app.bind('sequence').toClass(MySequence);
+  }
+
+  function givenProviders() {
+    class MyPassportStrategyProvider implements Provider<Strategy> {
+      constructor(
+        @inject(BindingKeys.Authentication.METADATA)
+        private metadata: AuthenticationMetadata,
+      ) {}
+      async value() : Promise<Strategy> {
+        if (this.metadata.strategy === 'BasicStrategy') {
+          return new BasicStrategy(this.verify);
+        } else {
+          return Promise.reject('configured strategy is not available');
+        }
```

Comment:
IMO, this code is too low-level and should be provided by the framework, so that application developers can stay on a higher level.

Ideally, as an app developer, I want to bind my custom strategy class, and the authentication should take care of the rest.

```ts
// in the app
class MyBasicStrategy extends BasicStrategy {
  constructor(@inject('repositories.user') private readonly userRepo) {
  }

  protected verify(username: string, password: string): Promise<UserProfile> {
    return userRepo.find(username, password);
  }
}

app.bind('BasicStrategy').toClass(MyBasicStrategy);
```

The `MyPassportStrategyProvider` can be bound by the `authentication` component then, and the code looking up different strategy implementations should stay inside this component.

Thoughts?

I guess this can be deferred as the next step after this patch is landed.